### ISSUE TITLE
image generation time  increased a lot on Mac M1

### ISSUE THREAD

**Issue by i3oc9i** [lstein/stable-diffusion#398]
**Describe your environment**
- GPU: mps
- VRAM: 128G
- CPU arch: arm
- OS:  macOS
- Python:  miniconda
- Branch:  development
- Commit: 720e5cd6513cd27e6d53feb6475dde20bd39841a

**Describe the bug**
after refactoring dream takes ~70% longer to generate images w.r.t. 1.13 release 

development : 1:26 sec
main : 0:50 sec

**To Reproduce**

git checkout development (720e5cd6513cd27e6d53feb6475dde20bd39841a)
```
dream> "test" -s50 -W576 -H896 -C7.5 -Ak_lms

100%|  50/50 [01:25<00:00,  1.71s/it]
Generating: 100%|  1/1 [01:26<00:00, 86.13s/it]

>> Usage stats:
>>   1 image(s) generated in 54.13s
>>   Max VRAM used for this generation: 0.00G
```
git checkout main (751283a2de81bee4bb571fbabe4adb19f1d85b97)

```
dream> "test" -s50 -W576 -H896 -C7.5 -Ak_lms

100%| 50/50 [00:50<00:00,  1.00s/it]
Generating: 100%| 1/1 [00:50<00:00, 50.87s/it]
>> Usage stats:
>>   1 image(s) generated in 50.95s
>>   Max VRAM used for this generation: 0.00G
```

**Environment**
```
pytorch                   1.12.1                 py3.10_0    pytorch
pytorch-lightning         1.6.5              pyhd8ed1ab_0    conda-forge
torch-fidelity            0.3.0                    pypi_0    pypi
torchdiffeq               0.2.3                    pypi_0    pypi
torchmetrics              0.9.3              pyhd8ed1ab_0    conda-forge
torchvision               0.13.1                py310_cpu    pytorch
```

**Comment by eduardoarinopelegrin** [comment]
The usage stats may also throw an odd value.
`1 image(s) generated in 50.95s` but right above
`Generating: 100%|  1/1 [01:26<00:00, 86.13s/it]`

**Comment by i3oc9i** [comment]
@eduardoarinopelegrin 

thank you for your remark, Indeed I forget to mention the discrepancy with the time reported in the stats.
but the **image realy takes 01:26 seconds to generate**, I double checked with a stopwatch at the same time.

**Comment by lstein** [comment]
@magnusviri , are you seeing slowdowns on your M1?

**Comment by eduardoarinopelegrin** [comment]
With 64 GB RAM, I can reproduce the issue. As a small note, the usage stats seem correct now.

git checkout development
```
dream> "test" -s50 -W576 -H896 -C7.5 -Ak_lms
Generating: 100%|| 1/1 [02:23<00:00, 143.55s/it]
>> Usage stats:
>>   1 image(s) generated in 143.94s
>>   Max VRAM used for this generation: 0.00G
```

git checkout main
```
dream> "test" -s50 -W576 -H896 -C7.5 -Ak_lms
Generating: 100%|| 1/1 [01:35<00:00, 95.73s/it]
>> Usage stats:
>>   1 image(s) generated in 95.84s
>>   Max VRAM used for this generation: 0.00G
```
Environment
```
pytorch                   1.12.1          cpu_py39h9b0ea23_0    conda-forge
pytorch-lightning         1.6.5              pyhd8ed1ab_0    conda-forge
torch-fidelity            0.3.0                    pypi_0    pypi
torchdiffeq               0.2.3                    pypi_0    pypi
torchmetrics              0.9.3              pyhd8ed1ab_0    conda-forge
torchvision               0.13.0          cpu_py39h982e834_0    conda-forge
```

**Comment by lstein** [comment]
I'm pretty sure the slowdowns on M1 hardware are due to the memory optimizations in attention.py that we instituted over the weekend. I don't have an M1 myself so can't check personally, but here's what you can do to help isolate the issue:
1. Check out the development branch `git checkout development; git pull` This will bring you up to date with all the recent changes.
2. Run your timing tests to confirm that problem still occurs.
3. Check out the version of attention.py just prior to the optimizations `git checkout bdb0651e ldm/modules/attention.py`
4. Run your timing tests again. Note whether the checkout has fixed them.
5. Return attention.py to its current state `git checkout HEAD ldm/modules/attention.py`

Much obliged for any help you can give in tracking this down.

**Comment by i3oc9i** [comment]
@lstein 

I guess you right, I executed the test procedure you ask for 

**1/ git checkout development; git pull** (eef788981cbed7c68ffd58b4eb22a2df2e59ae0b)
```
dream> "test" -s50 -W576 -H896 -C7.5 -Ak_lms

100%|| 50/50 [01:20<00:00,  1.61s/it]
Generating: 100%|| 1/1 [01:21<00:00, 81.05s/it]
>> Usage stats:
>>   1 image(s) generated in 81.25s
>>   Max VRAM used for this generation: 0.00G
```

**2/ git checkout bdb0651e ldm/modules/attention.py** 
```
dream> "test" -s50 -W576 -H896 -C7.5 -Ak_lms

100%| 50/50 [00:49<00:00,  1.01it/s]
Generating: 100%|| 1/1 [00:50<00:00, 50.01s/it]
>> Usage stats:
>>   1 image(s) generated in 50.16s
>>   Max VRAM used for this generation: 0.00G
```

so this is the same time I get on the 1.13 Relase

**Comment by eduardoarinopelegrin** [comment]
> I noticed similar behavior on m1 after changing size of image. When the first generation is like 384x384, second is 512x512, then speed drops (about 9 times slower). Terminal restart fixes it.

The bigger the image, the longer it takes to generate it (as there are more pixels). Although 9 times slower seems like a lot. Maybe using SWAP for 512x512 in your machine, but not for 384x384.

For me, `1 image(s) generated in 20.33s` for 512x512 and `1 image(s) generated in 29.81s` for 384x384. 50% increase in time for 77% increase in number of pixels.

**Comment by netsvetaev** [comment]
> > I noticed similar behavior on m1 after changing size of image. When the first generation is like 384x384, second is 512x512, then speed drops (about 9 times slower). Terminal restart fixes it.
> 
> The bigger the image, the longer it takes to generate it (as there are more pixels). Although 9 times slower seems like a lot. Maybe using Swap for 512x512 in your machine, but not for 384x384.
> 
> For me, `1 image(s) generated in 20.33s` for 512x512 and `1 image(s) generated in 29.81s` for 384x384. 50% increase in time for 77% increase in number of pixels. Much more reasonable.

Of course, but in my case it’s 9 times slower than 512 without changing size :-) I mean, if I just generate 512 it’s about 1:20 for 50 samples klms, but if I do that _after_ generating another size, for example 384, it takes 10-15 minutes. SOmetimes. Sometimes not.

**Comment by eduardoarinopelegrin** [comment]
@lstein 
ldm/modules/attention.py (Development branch)
```
        # attention, what we cannot get enough of, by halves
        sim[4:] = sim[4:].softmax(dim=-1)
        sim[:4] = sim[:4].softmax(dim=-1)

        sim = einsum('b i j, b j d -> b i d', sim, v)
        sim = rearrange(sim, '(b h) n d -> b n (h d)', h=h)
        return self.to_out(sim)
```
>> Usage stats:
>>   1 image(s) generated in 44.24s
>>   Max VRAM used for this generation: 0.00G

ldm/modules/attention.py (With this code instead)
```
        # attention, what we cannot get enough of
        attn = sim.softmax(dim=-1)
        out = einsum('b i j, b j d -> b i d', attn, v)
        out = rearrange(out, '(b h) n d -> b n (h d)', h=h)
        return self.to_out(out)
```
>> Usage stats:
>>   1 image(s) generated in 30.74s
>>   Max VRAM used for this generation: 0.00G

**Comment by lstein** [comment]
Two options here. The easier one is to detect if we are running on M1 by examing device.type=='mps', and choosing the old or new code depending on the results. The other is what you suggested, to interrogate VRAM and the image size, and choose whether to use the optimization if VRAM will be inadequate.

For the latter, we need to measure how much VRAM is used. Also, we don't actually know whether the memory optimization on CUDA results in any improvement in VRAM usage on M1, do we? I don't completely trust reports that "I couldn't run a large image before, and now I can" because VRAM is shared with other processes.

**Comment by eduardoarinopelegrin** [comment]
@lstein 
Yes, `if device.type=='mps'` is the easier solution. As for measuring VRAM, not sure how to do that in Windows/Linux, but in case some users with M1/M2 with low RAM (16? 8?) prefer the new code (that is, to trade speed for lower memory usage), since M1/M2 chips use unified memory, maybe we could do something like:
```
import psutil
use_halves = True if psutil.virtual_memory().available < 8 * 10**9 else False
```
then
```
use_mps = True if x.device.type == 'mps' else False # detect mps right before deleting x
del context, x
```
and finally
```
        # attention, what we cannot get enough of
        if use_mps and not use_halves:
            sim = sim.softmax(dim=-1)
        else:
            # by halves
            sim[4:] = sim[4:].softmax(dim=-1)
            sim[:4] = sim[:4].softmax(dim=-1)
        
        sim = einsum('b i j, b j d -> b i d', sim, v)
        sim = rearrange(sim, '(b h) n d -> b n (h d)', h=h)
        return self.to_out(sim)
```

To offer a bit more information about psutil, this is what `psutil.virtual_memory()` gives me (64 GB RAM). Something to note is, even though is says 10.7GB available, I can use 20GB extra (e.g. from python3.9 using 10GB to using 30GB in Activity Monitor) and still be fine (not use Swap). So maybe we should use `psutil.virtual_memory().total` instead of `psutil.virtual_memory().available` and assume users won't have many other things open at the time. This way, if a user has an M1/M2 with 8GB RAM, it would automatically use the 'by halves' version of the code.
```
% print(psutil.virtual_memory())
svmem(total=68719476736, available=10701914112, percent=84.4, used=4901261312, free=6921248768, active=4137091072, inactive=3623813120, wired=764170240)
```

From the psutil docs:  https://psutil.readthedocs.io/en/latest/

> **total**: total physical memory (exclusive swap).
> **available**: the memory that can be given instantly to processes without the system going into swap. This is calculated by summing different memory values depending on the platform and it is supposed to be used to monitor actual memory usage in a cross platform fashion.
> **used**: memory used, calculated differently depending on the platform and designed for informational purposes only. total - free does not necessarily match used.
> **free**: memory not being used at all (zeroed) that is readily available; note that this doesn’t reflect the actual memory available (use available instead). total - used does not necessarily match free.
> **active** (UNIX): memory currently in use or very recently used, and so it is in RAM.
> **inactive** (UNIX): memory that is marked as not used.
> **buffers** (Linux, BSD): cache for things like file system metadata.
> **cached** (Linux, BSD): cache for various things.
> **shared** (Linux, BSD): memory that may be simultaneously accessed by multiple processes.
> **slab** (Linux): in-kernel data structures cache.
> **wired** (BSD, macOS): memory that is marked to always stay in RAM. It is never moved to disk.

**Comment by eduardoarinopelegrin** [comment]
Just to confirm, the code works on my Mac.
[attention.py.zip](https://github.com/lstein/stable-diffusion/files/9498176/attention.py.zip)

**Comment by eduardoarinopelegrin** [comment]
> Also, we don't actually know whether the memory optimization on CUDA results in any improvement in VRAM usage on M1, do we?

Actually, I see worse performance memory-wise on my Mac with the new code.
`"ocean" -s50 -W512 -H512 -C7.5 -Ak_lms -F -S2381025714`
Old code
<img width="615" alt="Screenshot 2022-09-06 at 17 11 10" src="https://user-images.githubusercontent.com/50542132/188671281-5b56dc29-0cad-4b9a-bc25-fedaf8e12efa.png">
New code
<img width="604" alt="Screenshot 2022-09-06 at 17 12 18" src="https://user-images.githubusercontent.com/50542132/188671546-f268cd9d-efd9-43a8-a77e-10916717ce79.png">
Not sure if @i3oc9i  can reproduce this too?

**Comment by i3oc9i** [comment]
@eduardoarinopelegrin 
what do you want I test exactely, i'm not sure to understand what do you mean with old and new code ?

im on development branch eef788981cbed7c68ffd58b4eb22a2df2e59ae0b 

with this enviroment
```
Python 3.10.5

pytorch                   1.12.1                 py3.10_0    pytorch
pytorch-lightning         1.6.5              pyhd8ed1ab_0    conda-forge
torch-fidelity            0.3.0                    pypi_0    pypi
torchdiffeq               0.2.3                    pypi_0    pypi
torchmetrics              0.9.3              pyhd8ed1ab_0    conda-forge
torchvision               0.13.1                py310_cpu    pytorch
```

**Comment by eduardoarinopelegrin** [comment]
So, basically, this seems to be the problematic part of ldm/modules/attention.py in the Development branch.
```
        # attention, what we cannot get enough of, by halves
        sim[4:] = sim[4:].softmax(dim=-1)
        sim[:4] = sim[:4].softmax(dim=-1)

        sim = einsum('b i j, b j d -> b i d', sim, v)
        sim = rearrange(sim, '(b h) n d -> b n (h d)', h=h)
        return self.to_out(sim)
```
and this is the old version, which was faster.
```
# attention, what we cannot get enough of
        attn = sim.softmax(dim=-1)
        out = einsum('b i j, b j d -> b i d', attn, v)
        out = rearrange(out, '(b h) n d -> b n (h d)', h=h)
        return self.to_out(out)
```
lstein was asking it the first version, even though it was slower, at least used less RAM on Macs. I just tried and it actually was slower + used more RAM (instead of less). So no benefit at all.

I was asking if in your machine it also used more RAM, apart from being slower.

**Comment by eduardoarinopelegrin** [comment]
> @eduardoarinopelegrin what do you want I test exactely, i'm not sure to understand what do you mean with old and new code ?
> 
> in development branch [eef7889](https://github.com/lstein/stable-diffusion/commit/eef788981cbed7c68ffd58b4eb22a2df2e59ae0b) `"ocean" -s50 -W512 -H512 -C7.5 -Ak_lms -F -S2381025714 ` 50/50 [00:29<00:00, 1.67it/s]
> 
> in main branch [751283a](https://github.com/lstein/stable-diffusion/commit/751283a2de81bee4bb571fbabe4adb19f1d85b97) `"ocean" -s50 -W512 -H512 -C7.5 -Ak_lms -F -S2381025714` 50/50 [00:20<00:00, 2.46it/swith this enviroment
> 
> ```
> Python 3.10.5
> 
> pytorch                   1.12.1                 py3.10_0    pytorch
> pytorch-lightning         1.6.5              pyhd8ed1ab_0    conda-forge
> torch-fidelity            0.3.0                    pypi_0    pypi
> torchdiffeq               0.2.3                    pypi_0    pypi
> torchmetrics              0.9.3              pyhd8ed1ab_0    conda-forge
> torchvision               0.13.1                py310_cpu    pytorch
> ```

Yep, so simply check the memory usage as well on Activity Monitor. Do you see any differences between both branches?

**Comment by i3oc9i** [comment]
@eduardoarinopelegrin 

Development 15.4G   29sec
Main               11.65G 20Sec 

So development slower and more RAM

**Comment by i3oc9i** [comment]
Just to share that last nightly pythorch is on python 3.10 now

```
    ---------------------------|-----------------
    pytorch-1.13.0.dev20220905 |         py3.10_0        44.4 MB  pytorch-nightly
    torchaudio-0.13.0.dev20220905|        py310_cpu         5.6 MB  pytorch-nightly
    torchvision-0.14.0.dev20220906|        py310_cpu         6.4 MB  pytorch-nightly
    ------------------------------------------------------------
```
I will check if there is some improvment into a separate conda enviroment in order 
to avoid to break my official dream-dev test environment

**Comment by tildebyte** [comment]
@i3oc9i; Thanks for the leg work

I immediately note that 1.13 is "nightly" (and tagged with "dev"); I'd expect VRAM usage to drop darmatically once they cut a release. Whether it's below 1.11/1.12 remains to be seen :)

**Comment by glonlas** [comment]
> Actually, I see worse performance memory-wise on my Mac with the new code. 
I came here to share the same results. I switched to the development branch and I can relate the generation performance is painful to watch.

**Comment by tildebyte** [comment]
Let me rephrase: It's **highly** unlikely that the Torch team is going to **release** a huge memory usage regression.

**Comment by i3oc9i** [comment]
@glonlas 

I suggest you stay on main brach, developpment branch is a working in progress branch so things can be worse/broken, and there are actually other branches where performances  solutions are evaluated.

**Comment by Any-Winter-4079** [comment]
@i3oc9i The problem is some features such as `inpainting` are not available in the main branch.

Plus, if we want to add code (such as what I started and @glonlas took over, which is `upscaling` and` face restoration` for M1), I think it's better to develop it for the development branch, because they're refactoring code, so the main branch has a bit of an expiration date (and won't get all features implemented).

Oh, I just checked, and performance is back on the development branch! They brought this back
```
if device_type == 'mps':      #special case for M1 - disable neonsecret optimization
            sim = sim.softmax(dim=-1)
        else:
            sim[4:] = sim[4:].softmax(dim=-1)
            sim[:4] = sim[:4].softmax(dim=-1)
```

**Comment by tildebyte** [comment]
TL;DR - If you want something that works stably without issue, grab the release. If you want instability, but want to try out the latest features under development, `git checkout development`

HAVING SAID THAT

- there are some issues with release 1.13
- 'main' has an Mx fix on it (which it shouldn't)

so the repo isn't really in a "normal" state right now - hopefully things will be more "usual" after the next release

**Comment by lstein** [comment]
If you want to develop, add features, or propose bug fixes, use the development branch. The main branch will only change for urgent bug fixes.

**Comment by i3oc9i** [comment]
fixed in 1.14.1 release 9b28c65e4b0526956fd90ad86e2118536e4eaa9b
